File: Graph/LazyPrimMST.java
package Graph;

import StackAndQueue.Queue;
import sort.MinPQ;

/**
 * Created by sunkai on 2017/9/6
 */
public class LazyPrimMST {
    private boolean[] marked;
    private Queue<Edge> mst;
    private MinPQ<Edge> pq;
    public LazyPrimMST(EdgeWeightedGraph G){
        pq=new MinPQ<Edge>(G.E());
        marked=new boolean[G.V()];
        mst=new Queue<Edge>();
        visit(G,0);
        while(!pq.isEmpty()){
            Edge e=pq.delMin(); //获得权重最小的边
            int v=e.either();
            int w=e.other(v);
            if(marked[v]&&marked[w])    //跳过失败的边
                continue;;
            mst.enqueue(e); //将边添加到树中
            if(!marked[v])
                visit(G,v);
            if(!marked[w]){
                visit(G,w);
            }
        }
    }
    private void visit(EdgeWeightedGraph G,int v){
        //标记顶点V并将素有连接V且未被标记的顶点加入pq
        marked[v]=true;
        for(Edge e:G.adj(v)){
            if(!marked[e.other(v)])
                pq.insert(e);
        }
    }
    public Iterable<Edge> edges(){
        return mst;
    }
}

File: Algorithm_Exercises/Algorithm (4th edition)/Algorithm/Chapter4/MinimalSpanningTree/LazyPrimMST.cs


using Chapter1;
using Chapter2;
using System;
using System.Collections.Generic;

namespace Chapter4.MinimalSpanningTree
{
    public class LazyPrimMST:IMST
    {
        private bool[] marked;
        private Chapter1.Queue<IEdge> mst;
        private MinPQ<IEdge> pq;

        public LazyPrimMST(IEdgeWeightGraph g)
        {
            pq = new MinPQ<IEdge>();
            marked = new bool[g.V];
            mst = new Chapter1.Queue<IEdge>();
            Visit(g, 0);//假设G是联通的
            while(!pq.IsEmpty)
            {
                IEdge e = pq.Delete(); //找到权重最小的边
                int v = e.Either;
                int w = e.Other(v);
                if (marked[v] && marked[w]) //跳过失效的边
                    continue;
                mst.Enqueue(e); //将边添加到树中
                if (!marked[v])
                    Visit(g, v);
                if (!marked[w])
                    Visit(g, w);
            }
        }

        private void Visit(IEdgeWeightGraph g, int v)
        {
            //标记顶点v并将所有链接v和未被标记的顶点加入pq
            marked[v] = true;
            foreach (var edge in g.Adj(v))
            {
                if (!marked[edge.Other(v)])
                    pq.Insert(edge);
            }
        }

        public double Weight { get
            {
                double w = 0.0;
                foreach(var e in Edges())
                {
                    if (e != null)
                        w += e.Weight;
                }
                return w;
            } }

        public IEnumerable<IEdge> Edges() => mst;
    }
}
